ejb3/src/main/java/org/jboss/as/ejb3/component/stateful/StatefulSessionSynchronizationInterceptor.java

```diff
@@ -139,6 +154,22 @@ public Object processInvocation(final InterceptorContext context) throws Excepti
                         instance.getComponent().getCache().release(instance);
                     }
                 }
+                for(;;) {
+                    int state = invocationSyncState.get();
+                    if(state == SYNC_STATE_INVOCATION_IN_PROGRESS && invocationSyncState.compareAndSet(SYNC_STATE_INVOCATION_IN_PROGRESS, SYNC_STATE_NO_INVOCATION)) {
+                        break;
+                    } else if (state == SYNC_STATE_AFTER_COMPLETE_DELAYED) {
+                        try {
+                            //we know the transaction did not commit, as we are still active in the tx, this only happens through timeout
+                            handleAfterComplection(false, instance, toDiscard);
```

Comment:
Not sure if it is relevant but should we comment that other SessionSynchronization callbacks will have run already in the background (tm reaper) thread?